### PR TITLE
feat(gui): Persist local settings to config file

### PR DESCRIPTION
**Pull request by jhermsmeier**
This switches the settings to use a local configuration file, instead of `localStorage` to facilitate loading and manipulating the settings outside of a `window` context. Further, this decouples the settings from the applications state tracking redux store, and relocates models from `lib/shared/models` to `lib/gui/app/models`, as they weren't used in SDK code.

Previously added functionality to also respect a configuration file in the current working directory has been removed to simplify the mechanism and deconflict precedence.

Change-Type: minor
Connects To: https://github.com/resin-io/etcher/issues/2265

### CODE REVIEW

File: lib/gui/app/models/local-settings.js [hunk 1]
```diff
@@ -38,10 +48,19 @@ const settingsStorage = new Storage(LOCAL_STORAGE_SETTINGS_KEY)
  */
 const RCFILE = '.etcher.json'
 
+/**
+ * @summary Configuration file path
+ * @type {String}
+ * @constant
+ */
+const HOME_CONFIG_PATH = process.platform === 'win32'
+  ? path.join(USER_DATA_DIR, RCFILE)
```
Comment (lurch) [inline lib/gui/app/models/local-settings.js]: Is this special-casing of the filename still relevant on win32 now, given that `USER_DATA_DIR` will always end with `etcher` ?

**Comment by lurch** [comment]
Does this also fix #2342 ?

> and relocates models from lib/shared/models to lib/gui/app/models

Oooh, interesting - were they not used by any of the Etcher CLI code?

File: lib/gui/app/models/local-settings.js [hunk 2]
```diff
@@ -68,6 +87,35 @@ const readConfigFile = (filename) => {
   })
 }
 
+/**
+ * @summary Read a local .etcherrc file
```
Comment (lurch) [inline lib/gui/app/models/local-settings.js]: **Write** a local .etcherrc file? :wink:

File: lib/gui/app/modules/image-writer.js [hunk 3]
```diff
@@ -313,6 +313,7 @@ exports.flash = (image, drives) => {
   analytics.logEvent('Flash', analyticsData)
 
   return exports.performWrite(image, drives, (state) => {
+    console.log('performWriate.onProgress', state)
```
Comment (zvin) [inline lib/gui/app/modules/image-writer.js]: performWrite

File: lib/gui/app/models/settings.js [hunk 4]
```diff
@@ -156,9 +157,16 @@ exports.set = (key, value) => {
     }))
   }
 
+  const previousValue = settings[key]
+
   settings[key] = value
 
   return localSettings.writeAll(settings)
+    .catch((error) => {
```
Comment (zvin) [inline lib/gui/app/models/settings.js]: NIT: `writeAll` returns a Bluebird promise so you can use `.tapCatch`

Comment (jhermsmeier) [reply lib/gui/app/models/settings.js]: Aye, added!

File: lib/gui/app/models/local-settings.js (hunk 2 above)
Comment (jhermsmeier) [reply lib/gui/app/models/local-settings.js]: Yup, mended!

File: lib/gui/app/models/local-settings.js (hunk 1 above)
Comment (jhermsmeier) [reply lib/gui/app/models/local-settings.js]: Ah, good catch – this is obsolete now, as is `RCFILE`

File: lib/gui/app/models/local-settings.js [hunk 5]
```diff
@@ -16,32 +16,42 @@
...
+
+/**
+ * @summary Configuration file path
+ * @type {String}
+ * @constant
+ */
+const CONFIG_PATH = path.join(USER_DATA_DIR, 'config.json')
 
 /**
  * @summary Read a local .etcherrc file
```
Comment (lurch) [inline lib/gui/app/models/local-settings.js]: Perhaps this should be changed to `Read the local configuration file` to match the jsdoc on line 82?

File: lib/gui/app/models/settings.js (hunk 4 above)
Comment (jhermsmeier) [reply lib/gui/app/models/settings.js]: Had to revert this, as this method isn't a thing with the version of Bluebird we're using atm yet

File: lib/gui/app/models/local-settings.js (hunk 5 above)
Comment (jhermsmeier) [reply lib/gui/app/models/local-settings.js]: Done

File: lib/gui/app/models/store.js [hunk 6]
```diff
@@ -550,3 +508,34 @@ module.exports = _.merge(redux.createStore(storeReducer, DEFAULT_STATE), {
...
+    const nextState = module.exports.getState()
+    if (!_.isEqual(nextState, currentState)) {
+      currentState = nextState
+      onChange(currentState)
+    }
+  }
+
+  const unsubscribe = module.exports.subscribe(changeHandler)
+  changeHandler()
+  return unsubscribe
```
Comment (Shou) [inline lib/gui/app/models/store.js]: Do we need the `changeHandler()` between `const unsubscribe` and `return unsubscribe` or could we consolidate the const-setting with the return statement? Either works for me I'm just curious if I've missed something subtle.

**Pull request merged**